Migrate to Swift 3

dusek opened the pull request:
Also migrates to Nimble 6.1.

The only backwards-incompatible change is I made the `Type` and its `static let` members `public`, not `open` – anyone who subclassed `Type` now cannot, but I made it on the assumption that subclassing `Type` was never intended and is not desired.

All tests pass except for test message failure, which is due to missing/superfluous double quotes around some strings inside json dictionary in one of the compared messages:

- [Got failure message: "expected to include <["name": Pikachu]>, got <["no": 25, "name": "Pikachu"]>", but expected "expected to include <["name": Pikachu]>, got <["no": 25, "name": Pikachu]>"](https://github.com/giginet/JSONMatcher/blob/master/Tests/JSONMatcher/BeJSONIncludingTests.swift#L29)
- [Got failure message: "expected to not include <["name": Mew]>, got <["no": 25, "name": "Pikachu"]>", but expected "expected to not include <["name": Mew]>, got <["no": 25, "name": Pikachu]>"](https://github.com/giginet/JSONMatcher/blob/master/Tests/JSONMatcher/BeJSONIncludingTests.swift#L32)

I did not solve these, as I think these require a deeper understanding of how (and why) the error messages are generated.

dusek committed:
Xcode Convert to Swift 3

dusek committed:
Upgrade Nimble (v4.0.1 -> v6.1.0)

dusek committed:
Fix location of Nimble's utils.swift

dusek committed:
Fix remaining build failures

dusek committed:
Fix remaining test build failures

dusek committed:
Travis: upgrade Xcode image (7.3 -> 8.3)

dusek committed:
Fix Xcode automatic migration

dusek committed:
Travis: do not install swiftlint

swiftlint is apparently already installed on Travis CI Xcode 8.3
image.

dusek committed:
Add SWIFT_VERSION for tvOS and macOS projects too

dusek committed:
Travis: run `pod repo update` before running linting podspec

dusek committed:
Adjust macOS deployment target to match Nimble's

dusek committed:
Add .swift-version file for CocoaPods

dusek commented:
Please note the failure of `pod lib lint` is due to a Swift compiler crash [SR-4549](https://bugs.swift.org/browse/SR-4549) I just reported.

giginet commented:
@dusek  Thank you very much.
I'll review it in a few days.

giginet commented on .swift-version:
```diff
@@ -0,0 +1 @@
+3.0
```
Could you upgrade to 3.1

dusek replied:
I could, I am just not sure that actually does anything. I have not found docs for `.swift-version` meaning, and even Nimble (and Alamofire) seem to specify `3.0` even though the latest version has been updated for Swift 3.1. So unless you are sure that specifying `3.1` there is the right thing to do, I would leave it at `3.0.`. I no longer am able to try things with Swift 3.0, so it would be probably hard to test with that release of the language.

dusek replied:
I also found out that Xcode's automatic migration puts "SWIFT_VERSION = 3.0" into the xcodeproj file (even though it was Swift 3.1 version doing the migration), so given that also the popular Swift projects updated to 3.1 are also stating "3.0" in their `.swift-version`, I am inclined to rather leave .swift-version at "3.0".

So I left it at "3.0" to be safe.

giginet commented on .travis.yml:
```diff
@@ -1,15 +1,13 @@
-      osx_image: xcode7.3
+      osx_image: xcode8.3
 script: 
+  - pod repo update
```
Is this necessary?

dusek replied:
I no longer remember :-) I think it was due to the fact that the latest cocoapods (present on Travis CI) does not do the update automatically during other commands, and without the update it was not finding the latest Nimble version in mentioned in JSONMatcher's cocoapods spec. I will check once again how it behaves without the update.

dusek replied:
I tried without `pod repo update` and it works, so I removed it. Probably this is a problem only when JSONMatcher quickly upgrade to a new version of a library, in the initial days probably Travis might not have the pod repo updated to include the newest version.

giginet commented on Sources/JSONMatcher/Builder.swift:
```diff
@@ -14,13 +14,13 @@ struct Builder {
     }
 
-    private func buildRawJSONElement<T>(object: T) -> BaseElementType {
+    fileprivate func buildRawJSONElement<T>(_ object: T) -> BaseElementType {
```
They should be `private`

dusek replied:
I guess so, I will correct this, and try to "privatize" all other functions and see how far I can go with that without breaking the code.

dusek replied:
I made all `fileprivate` methods and properties into `private`, and things still work.

giginet commented on Tests/JSONMatcher/BaseTestCase.swift:
```diff
@@ -2,10 +2,10 @@ import Foundation
-        let path = bundle.pathForResource(JSONName, ofType: "json")!
-        let jsonData = NSData(contentsOfFile: path)
-        return String(data: jsonData!, encoding: NSUTF8StringEncoding)!
+    func loadJSONFile(_ JSONName: String) -> String {
```
`jsonName`
according to API design guideline

dusek replied:
indeed, will fix

dusek replied:
Fixed this, and also 2 other similar cases.

giginet reviewed (approved):
Sorry for waiting you @dusek  🙇 

I reviewed. 
I think most of `fileprivate` could be replaced to `private`.

giginet commented:
And some tests are failing.
https://travis-ci.org/giginet/JSONMatcher/builds/220218021

Could you also fix this?

dusek committed:
Remove `pod repo update` from Travis CI

No longer seems necessary.

dusek committed:
Review comment: adhere to Swift API guidelines

dusek committed:
Change `fileprivate` to `private`

dusek committed:
Adhere to Swift API guidelines

dusek committed:
Workaround for SR-4549

This did not affect Run and Test, but did affect Profile and
Archive (release builds).

Tested with Xcode 8.3.2 (Apple Swift version 3.1 (swiftlang-802.0.53
clang-802.0.42))

dusek committed:
Improve reliability of (fix) run_tests

Previously run_tests exit status did not fail if the tests
failed. Additional sh settings are needed to make it fail when our
unit tests fail.

Solution found using based on what fastlane's `scan` command does.

dusek committed:
Fix test failures due to change in stringification

In Swift 3, `String(describing: ...)` formats `["a": "b"]` as such,
however formats `"a"` as simply `a`, i.e. without quotes. This leads
to discrepancies when Nimble stringifies actual value in the form of a
`Dictionary` (i.e. leaf `String`s have quotes), and when JSONMatcher
stringifies the expected value in the form of
DictionaryElement (i.e. leaf `String`s, in this case `StringElement`
having no quotes). The solution to ensure consitency between the two
is therefore to stringify `StringElement`s with quotes around the
actual string.

dusek committed:
Remove swiftlint warnings [colon]

dusek committed:
Remove swiftlint warnings [trailing_comma]

dusek commented:
@giginet Thanks for the review. I addressed your comments in commits and/or commented on them above. Also:

- I managed to find a way to circumvent the compiler crash by reformulation of the code. So now this part of Travis tests (`pod lib lint`) no longer fails.
- I fixed the `run_tests` to actually exit non-zero when unit tests fail.
- I fixed the failing unit tests.
- (The above three items mean that now the Travis tests are green and rightfully so.)
- I fixed most of the `swiftlint` warnings.

Please review.

dusek committed:
Fix build failure

giginet reviewed (approved):
@dusek LGTM

Sorry for waiting for you such a long time 😢 

This product is no longer maintained.
If you wish do you become a contributor?
